src/libhirte/common/config.c

```diff
@@ -0,0 +1,175 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "libhirte/hashmap/hashmap.h"
+#include "libhirte/ini/ini.h"
+
+#include "config.h"
+
+
+/*
+ * The application configuration object.
+ */
+struct hashmap *config;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+int option_compare(const void *a, const void *b, void *udata) { // NOLINT
+        const struct config_option *option_a = a;
+        const struct config_option *option_b = b;
+        return strcmp(option_a->name, option_b->name);
+}
+#pragma GCC diagnostic pop
+
+uint64_t option_hash(const void *item, uint64_t seed0, uint64_t seed1) {
+        const struct config_option *option_a = item;
+        return hashmap_sip(option_a->name, strlen(option_a->name), seed0, seed1);
+}
+
+int config_initialize() {
+        config = hashmap_new(sizeof(struct config_option), 0, 0, 0, option_hash, option_compare, NULL, NULL);
+        if (config == NULL) {
+                // error during hashmap initialization
+                return ENOMEM;
+        }
+        return 0;
+}
+
+void config_dispose() {
+        hashmap_free(config);
+}
+
+int config_load_complete_cofiguration(
+                const char *default_config_file,
+                const char *custom_config_file,
+                const char *custom_config_directory) {
+        int result = 0;
+
+        if (default_config_file != NULL) {
+                result = config_load_from_file(default_config_file);
+                if (result != 0) {
+                        return result;
+                }
+        }
+
+        if (custom_config_file != NULL) {
+                result = config_load_from_file(custom_config_file);
+                if (result != 0) {
+                        return result;
+                }
+        }
+
+        if (custom_config_directory != NULL) {
+                // TODO: https://github.com/containers/hirte/issues/148
+        }
+
+        result = config_load_from_env();
+
+        return result;
+}
+
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+/*
+ * Handle used to store option parsed from a file into a configuration.
+ *
+ * Returns 1 on success, 0 on error as per ini library requirements.
+ */
+static int parsing_handler(void *user, const char *section, const char *name, const char *value) { // NOLINT
+        if (config_set_value(name, value) != 0) {
+                return 0;
+        }
+        return 1;
+}
+#pragma GCC diagnostic pop
+
+int config_load_from_file(const char *config_file) {
+        assert(config_file != NULL);
```

Comment:
```suggestion
        if (config_file == NULL) {
                return -EINVAL;
        }
```
I think in this case we could just return an error code and let the caller handle it. Could you change this also in all other places?

Reply:
Fixed